### ISSUE TITLE
Python - "Connection refused" exception during wait and repeated script execution

### ISSUE THREAD

**Issue by freshquiz** [SeleniumHQ/selenium#2088]
## Meta -
OS:  Ubuntu 16.04 x64

Selenium Version:  2.53.2 (Python 3)

Browser:  Google Chome and Firefox

Browser Version:  46.0 and 50.0.2661.94

## Expected Behavior -
No "[Errno 111] Connection refused" exception to be raised.
## Actual Behavior -
"[Errno 111] Connection refused" exception is raised.
## Steps to reproduce -

I can't provide a small example because the issue doesn't consistently occur, it's quite sporadic.

The best I can provide is a stack trace:

```
  File "my_client_code.py", line 21, in _record_id_present
    'return !!$(".tab_window_form").attr("data-record_id")')
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/remote/webdriver.py", line 461, in execute_script
    {'script': script, 'args':converted_args})['value']
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/remote/webdriver.py", line 231, in execute
    response = self.command_executor.execute(driver_command, params)
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/remote/remote_connection.py", line 395, in execute
    return self._request(command_info[0], url, body=data)
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/remote/remote_connection.py", line 425, in _request
    self._conn.request(method, parsed_url.path, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1106, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1151, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.5/http/client.py", line 1102, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/usr/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/usr/lib/python3.5/http/client.py", line 849, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/usr/lib/python3.5/socket.py", line 711, in create_connection
    raise err
  File "/usr/lib/python3.5/socket.py", line 702, in create_connection
    sock.connect(sa)
[Errno 111] Connection refused
```

So I'm calling `WebDriver.execute_script()` (checking if an element has a certain attribute value), from a `WebDriverWait.until()`.

It affects both Chrome and Firefox, so I'm guessing that rules out issues with the web drivers and places the blame on Selenium, somehow.

If it helps, the issue usually arises when I max out of my CPU cores with parallel processes of differing client code (which all do the same wait condition, at various stages, multiple times).

**Comment by freshquiz** [comment]
With this:
`File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/remote/remote_connection.py", line 425, in _request`
... the connection refused error arises when a connection tries to go through to the wrong port.

I.e. Prior connections are made to the correct port (where the driver server socket is listening) and then one is made to the wrong port. 

As to why the port suddenly changes, I do not know.
I will try and investigate that myself now...

**Comment by freshquiz** [comment]
Fault was my own.

I was caching waiter objects by wait time and not by (browser, wait time) and that causing connections to be made to the wrong driver socket addresses (which had already been closed).

**Comment by ratatatatat** [comment]
Hi, I'm having a similar problem, could you elaborate on what you did to fix the issue. The following block of code is producing the error is:
`	    element = WebDriverWait(driver, 20).until(
	        EC.presence_of_element_located((By.XPATH, '//*[@id="rso"]/div/div/div[1]/div/div/h3/a'))
	    )
`
The error:
`Traceback (most recent call last):
  File "NasTA.py", line 179, in <module>
    main()
  File "NasTA.py", line 174, in main
    driver.close()
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 537, in close
    self.execute(Command.CLOSE)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 250, in execute
    response = self.command_executor.execute(driver_command, params)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/remote_connection.py", line 464, in execute
    return self._request(command_info[0], url, body=data)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/remote_connection.py", line 487, in _request
    self._conn.request(method, parsed_url.path, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1057, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1097, in _send_request
    self.endheaders(body)
  File "/usr/lib/python2.7/httplib.py", line 1053, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 897, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 859, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 836, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python2.7/socket.py", line 575, in create_connection
    raise err
socket.error: [Errno 111] Connection refused
`

**Comment by begincalendar** [comment]
@ratatatatat I was caching the wait objects in a poor way, such that when a browser X was closed and I thought I was waiting on browser Y (which would still be open), it would actually be waiting on browser X (whose socket is no longer open, hence the error).

Just make sure the browser your `driver` points to (at the time of that call), is still open.

Otherwise it's probably another issue.

**Comment by deals4kr** [comment]
I am a beginner, i am getting same error and stack trace is here: Kindly advice what can be done to avoid this error?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 1318, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py", line 936, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/socket.py", line 724, in create_connection
    raise err
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/JyothiKondaReddy/Documents/workspace/PySelenium/pySelPack/pymodule.py", line 10, in <module>
    browser = webdriver.Remote(desired_capabilities=webdriver.DesiredCapabilities.SAFARI)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py", line 154, in __init__
    self.start_session(desired_capabilities, browser_profile)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py", line 243, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py", line 310, in execute
    response = self.command_executor.execute(driver_command, params)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/selenium/webdriver/remote/remote_connection.py", line 466, in execute
    return self._request(command_info[0], url, body=data)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/selenium/webdriver/remote/remote_connection.py", line 528, in _request
    resp = opener.open(request, timeout=self._timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 1346, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 61] Connection refused>
--------------------------------------------------------------------------

My code is 
----------
'''
loginsafari.py
'''
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
from selenium import webdriver
import urllib

browser = webdriver.Remote(desired_capabilities=webdriver.DesiredCapabilities.SAFARI)
#browser=webdriver.Safari()
browser.get('http://www.google.com')

**Comment by lmtierney** [comment]
@deals4kr please don't comment on old closed issues. If you have an issue please open a new one with all the data requested from the new issue template

**Comment by deals4kr** [comment]
Sure. Sorry for that, I am new user just joined.
